readme.md
[![Build Status](https://travis-ci.com/mapbox/pagerduty.svg?branch=master)](https://travis-ci.com/mapbox/pagerduty)

# pagerduty

A Node.js SDK for the PagerDuty v2 API

## API

First, create a new PagerDuty SDK object. A PagerDuty API token that will be used by default for all methods can be included by specifying the `pagerDutyToken` option.

**Parameters**

-   `options.pagerDutyToken` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** PagerDuty access token
-   `options.timeout` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)?** number of milliseconds used for read and connection timouets - defaults to 10 seconds if not specified


**Example**

```javascript
const PagerDuty = require('@mapbox/pagerduty');
const pd = new PagerDuty({
  pagerDutyToken: process.env.API_TOKEN
});

pd.get({
  path: 'services'
  ...
```

<!-- Generated, in part, by documentation.js. Update this documentation by updating the source code. -->

- [get](#get)
- [post](#post)
- [put](#put)
- [delete](#delete)

### get

Make a GET request to the PagerDuty API. You can choose to receive a specific
response body property by specifying an `options.key`; otherwise, the entire
response body will be returned. The PagerDuty access token can be provided
explicitly.

**Parameters**

-   `options.path` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** URL path and query
-   `options.key` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)?** Response body property to return
-   `options.token` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)?** PagerDuty access token
-   `options.timeout` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)?** number of milliseconds used for read and connection timouets

**Example**

```javascript
const PagerDuty = require('@mapbox/pagerduty');
const pd = new PagerDuty();
const id = 'PPPPPPA';
pd.get({
  path: `users/${id}`,
  key: 'user'
}, (err, res) => {
  if (err) throw err;
  console.log(`Successfully fetched ${res[0].name} with ID ${id}.`);
});
```

### post

Make a POST request to the PagerDuty API. The PagerDuty access token can be
provided explicitly.

**Parameters**

-   `options.path` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** URL path and query
-   `options.body` **[object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object)** POST request body
-   `options.headers` **[object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object)?** Additional headers, excluding `Accept`
    and `Authorization`
-   `options.token` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)?** PagerDuty access token
-   `options.timeout` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)?** number of milliseconds used for read and connection timouets

**Example**

```javascript
const PagerDuty = require('@mapbox/pagerduty');
const pd = new PagerDuty();
pd.post({
  path: 'users',
  body: {
    user: {
      type: 'user',
      name: 'Dev Null',
      email: 'devnull@company.com',
      time_zone: 'America/New_York',
      color: 'blue-violet',
      role: 'user'
    }
  },
  headers: {
    From: 'anotherdevnull@company.com'
  }
}, (err, res) => {
  if (err) throw err;
  console.log(`Created ${res.body.user.summary} with ID ${res.body.user.id}.`);
});
```

### put

Make a PUT request to the PagerDuty API. The PagerDuty access token can be
provided explicitly.

**Parameters**

-   `options.path` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** URL path and query
-   `options.body` **[object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object)** PUT request body
-   `options.headers` **[object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object)?** Additional headers, excluding `Accept`
    and `Authorization`
-   `options.token` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)?** PagerDuty access token
-   `options.timeout` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)?** number of milliseconds used for read and connection timouets

**Example**

```javascript
const PagerDuty = require('@mapbox/pagerduty');
const pd = new PagerDuty();
const id = 'PPPPPPA';
pd.put({
  path: `users/${id}`,
  body: {
    user: {
      job_title: 'Software Engineer'
    }
  }
}, (err, res) => {
  if (err) throw err;
  console.log(`Successfully updated ${res.body.user.name}'s user description to ${res.body.user.description}`);
});
```

### delete

Make a DELETE request to the PagerDuty API. The PagerDuty access token can
be provided explicitly.

**Parameters**

-   `options.path` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)** URL path and query
-   `options.token` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)?** PagerDuty access token
-   `options.timeout` **[string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String)?** number of milliseconds used for read and connection timouets

**Example**

```javascript
const PagerDuty = require('@mapbox/pagerduty');
const pd = new PagerDuty();
const id = 'PPPPPPA';
pd.delete({
  path: `users/${id}`
}, (err, res) => {
  if (err) throw err;
  console.log(`Successfully deleted user ID ${id}.`);
});
```


changelog.md
### 4.0.0

- Sets a default timeout of 10 seconds and supports custom timeout option for both the client and methods

### 3.0.0

- Module now exports a PagerDuty() class
- Exported class can be configured with an access token

### 2.0.0

- Node 6 upgrade

### 1.0.0

- Initial release of `get`, `post`, `put`, and `delete` methods


test/index.test.js
'use strict';

/* Store and set environment variables */
const originalTEST = process.env.TEST;
const originalCustomPagerDutyToken = process.env.CustomPagerDutyToken;
process.env.TEST = true;
process.env.CustomPagerDutyToken = 'test_token';

/* Initialize constants */
const fs = require('fs');
const PagerDuty = require(`${__dirname}/../index`);
const pd = new PagerDuty();
const test = require('tape');
let server, app;

/* Fixtures */
const createOverride = JSON.parse(fs.readFileSync(`${__dirname}/fixtures/createOverride.json`));
const createUser = JSON.parse(fs.readFileSync(`${__dirname}/fixtures/createUser.json`));
const getEscalationPolicy = JSON.parse(fs.readFileSync(`${__dirname}/fixtures/getEscalationPolicy.json`));
const listIncidentLogEntries = JSON.parse(fs.readFileSync(`${__dirname}/fixtures/listIncidentLogEntries.json`));
const listIncidentNotes = JSON.parse(fs.readFileSync(`${__dirname}/fixtures/listIncidentNotes.json`));
const listIncidents = JSON.parse(fs.readFileSync(`${__dirname}/fixtures/listIncidents.json`));
const listSchedules = JSON.parse(fs.readFileSync(`${__dirname}/fixtures/listSchedules.json`));
const listUsers = JSON.parse(fs.readFileSync(`${__dirname}/fixtures/listUsers.json`));
const updateUser = JSON.parse(fs.readFileSync(`${__dirname}/fixtures/updateUser.json`));

test('[pagerduty]', (t) => {
  t.test('[pagerduty] setup mock server', (assert) => {
    app = require('express')();

    app.post('/schedules/PPPPPPA/overrides', (req, res) => {
      assert.equal(req.url, '/schedules/PPPPPPA/overrides', 'expected req.url');
      res.status(201).send(createOverride);
    });

    app.post('/users', (req, res) => {
      assert.equal(req.url, '/users', 'expected req.url');
      assert.ok(req.headers.accept && req.headers.authorization && req.headers.from, 'combined default & user-provided headers');
      res.status(201).send(createUser);
    });

    app.get('/escalation_policies/PPPPPPA', (req, res) => {
      assert.equal(req.url, '/escalation_policies/PPPPPPA?offset=0&total=true&limit=100', 'expected req.url');
      res.status(200).send(getEscalationPolicy);
    });

    app.get('/escalation_policies/PPPPPPA/log_entries', (req, res) => {
      assert.equal(req.url, '/escalation_policies/PPPPPPA/log_entries?offset=0&total=true&limit=100', 'expected req.url');
      res.status(200).send(listIncidentLogEntries);
    });

    app.get('/escalation_policies/PPPPPPA/notes', (req, res) => {
      assert.equal(req.url, '/escalation_policies/PPPPPPA/notes?offset=0&total=true&limit=100', 'expected req.url');
      res.status(200).send(listIncidentNotes);
    });

    app.get('/incidents', (req, res) => {
      assert.equal(req.url, '/incidents?since=2017-04-04T13%3A00%3A00Z&time_zone=UTC&offset=0&total=true&limit=100', 'expected req.url');
      res.status(200).send(listIncidents);
    });

    app.get('/schedules', (req, res) => {
      assert.equal(req.url, '/schedules?query=cat&offset=0&total=true&limit=100', 'expected req.url');
      res.status(200).send(listSchedules);
    });

    app.get('/users', (req, res) => {
      assert.equal(req.url, '/users?query=Dev%20Null&offset=0&total=true&limit=100', 'expected req.url');
      res.status(200).send(listUsers);
    });

    app.put('/users/PPPPPPA', (req, res) => {
      assert.equal(req.url, '/users/PPPPPPA', 'expected req.url');
      res.status(200).send(updateUser);
    });

    app.delete('/users/PPPPPPA', (req, res) => {
      assert.equal(req.url, '/users/PPPPPPA', 'expected req.url');
      res.status(204).send('');
    });

    assert.end();
  });

  t.test('[pagerduty] start server', (assert) => {
    server = app.listen(3333, () => {
      assert.end();
    });
  });

  t.test('[pagerduty] [post] createOverride', (assert) => {
    pd.post({
      path: 'schedules/PPPPPPA/overrides',
      body: {
        override: {
          start: '2017-04-06T12:00:00+00:00',
          end: '2017-04-07T12:00:00+00:00',
          user: {
            id: 'PPPPPPA',
            type: 'user_reference',
            summary: 'Dev Null',
            self: 'https://api.pagerduty.com/users/PPPPPPA',
            html_url: 'https://company.pagerduty.com/users/PPPPPPA'
          }
        }
      }
    }, (err, res) => {
      assert.ifError(err, 'should not error');
      assert.deepEqual(res.body, createOverride);
      assert.end();
    });
  });

  t.test('[pagerduty] [post] createUser', (assert) => {
    pd.post({
      path: 'users',
      body: {
        user: {
          type: 'user',
          name: 'Dev Null',
          email: 'devnull@company.com',
          time_zone: 'America/New_York',
          color: 'blue-violet',
          role: 'user'
        }
      },
      headers: {
        From: 'anotherdevnull@company.com'
      }
    }, (err, res) => {
      assert.ifError(err, 'should not error');
      assert.deepEqual(res.body, createUser);
      assert.end();
    });
  });

  t.test('[pagerduty] [get] getEscalationPolicy', (assert) => {
    pd.get({
      key: 'escalation_policy',
      path: 'escalation_policies/PPPPPPA'
    }, (err, res) => {
      assert.ifError(err, 'should not error');
      assert.deepEqual(res, [getEscalationPolicy.escalation_policy], 'expected pd.get response');
      assert.end();
    });
  });

  t.test('[pagerduty] [get] listIncidentLogEntries', (assert) => {
    pd.get({
      key: 'log_entries',
      path: 'escalation_policies/PPPPPPA/log_entries'
    }, (err, res) => {
      assert.ifError(err, 'should not error');
      assert.deepEqual(res, listIncidentLogEntries.log_entries, 'expected pd.get response');
      assert.end();
    });
  });

  t.test('[pagerduty] [get] listIncidentNotes', (assert) => {
    pd.get({
      key: 'notes',
      path: 'escalation_policies/PPPPPPA/notes'
    }, (err, res) => {
      assert.ifError(err, 'should not error');
      assert.deepEqual(res, listIncidentNotes.notes, 'expected pd.get response');
      assert.end();
    });
  });

  t.test('[pagerduty] [get] listIncidents', (assert) => {
    pd.get({
      key: 'incidents',
      path: 'incidents?since=2017-04-04T13%3A00%3A00Z&time_zone=UTC'
    }, (err, res) => {
      assert.ifError(err, 'should not error');
      assert.deepEqual(res, listIncidents.incidents, 'expected pd.get response');
      assert.end();
    });
  });

  t.test('[pagerduty] [get] listSchedules', (assert) => {
    pd.get({
      key: 'schedules',
      path: 'schedules?query=cat'
    }, (err, res) => {
      assert.ifError(err, 'should not error');
      assert.deepEqual(res, listSchedules.schedules, 'expected pd.get response');
      assert.end();
    });
  });

  t.test('[pagerduty] [get] listUsers', (assert) => {
    pd.get({
      key: 'users',
      path: 'users?query=Dev%20Null'
    }, (err, res) => {
      assert.ifError(err, 'should not error');
      assert.deepEqual(res, listUsers.users, 'expected pd.get response');
      assert.end();
    });
  });

  t.test('[pagerduty] [get] listUsers, no key', (assert) => {
    pd.get({
      path: 'users?query=Dev%20Null'
    }, (err, res) => {
      assert.ifError(err, 'should not error');
      assert.deepEqual(res, [listUsers], 'expected pd.get response');
      assert.end();
    });
  });

  t.test('[pagerduty] [put] updateUsers', (assert) => {
    pd.put({
      path: 'users/PPPPPPA',
      body: {
        user: {
          description: 'Software Engineer'
        }
      }
    }, (err, res) => {
      assert.ifError(err, 'should not error');
      assert.deepEqual(res.body, updateUser, 'expected pd.put response');
      assert.end();
    });
  });

  t.test('[pagerduty] [delete] deleteUser delete', (assert) => {
    pd.delete({
      path: 'users/PPPPPPA'
    }, (err, res) => {
      assert.ifError(err, 'should not error');
      assert.deepEqual(res.body, '');
      assert.end();
    });
  });

  t.test('[pagerduty] cleanup', (assert) => {
    server.close();
    process.env.TEST = originalTEST;
    process.env.CustomPagerDutyToken = originalCustomPagerDutyToken;
    assert.end();
  });

  t.end();
});


test/utils.test.js
'use strict';

const file = require(`${__dirname}/../lib/utils`);
const test = require('tape');

test('[utils] [updateUrl]', (t) => {
  t.equal(file.updateUrl('http://localhost:3333/users?query=Test%20User', 10), 'http://localhost:3333/users?query=Test%20User&offset=10&total=true&limit=100', 'no search parameters provided');
  t.equal(file.updateUrl('http://localhost:3333/users?query=Test%20User&offset=5', null), 'http://localhost:3333/users?query=Test%20User&offset=5&total=true&limit=100', 'offset provided');
  t.equal(file.updateUrl('http://localhost:3333/users?query=Test%20User&total=false', 10), 'http://localhost:3333/users?query=Test%20User&total=false&offset=10&limit=100', 'total provided');
  t.equal(file.updateUrl('http://localhost:3333/users?query=Test%20User&limit=100', 10), 'http://localhost:3333/users?query=Test%20User&limit=100&offset=10&total=true', 'offset provided');
  t.equal(file.updateUrl('http://localhost:3333/users?query=Test%20User&limit=100&total=false&offset=5', null), 'http://localhost:3333/users?query=Test%20User&limit=100&total=false&offset=5', 'all search parameters provided');
  t.end();
});


test/pagerduty.test.js
'use strict';
process.env.TEST = true;

const fs = require('fs');
const test = require('tape');
const PagerDuty = require('../lib/pagerduty');
const pd = new PagerDuty({ pagerDutyToken: 'fakeaccesstoken' });

let server, app;

/* Fixtures */
const createOverride = JSON.parse(fs.readFileSync(`${__dirname}/fixtures/createOverride.json`));
const createUser = JSON.parse(fs.readFileSync(`${__dirname}/fixtures/createUser.json`));
const getEscalationPolicy = JSON.parse(fs.readFileSync(`${__dirname}/fixtures/getEscalationPolicy.json`));
const listIncidentLogEntries = JSON.parse(fs.readFileSync(`${__dirname}/fixtures/listIncidentLogEntries.json`));
const listIncidentNotes = JSON.parse(fs.readFileSync(`${__dirname}/fixtures/listIncidentNotes.json`));
const listIncidents = JSON.parse(fs.readFileSync(`${__dirname}/fixtures/listIncidents.json`));
const listSchedules = JSON.parse(fs.readFileSync(`${__dirname}/fixtures/listSchedules.json`));
const listUsers = JSON.parse(fs.readFileSync(`${__dirname}/fixtures/listUsers.json`));
const updateUser = JSON.parse(fs.readFileSync(`${__dirname}/fixtures/updateUser.json`));

test('[pagerduty]', (t) => {
  t.test('[pagerduty] setup mock server', (assert) => {
    app = require('express')();

    app.post('/schedules/PPPPPPA/overrides', (req, res) => {
      assert.equal(req.url, '/schedules/PPPPPPA/overrides', 'expected req.url');
      res.status(201).send(createOverride);
    });

    app.post('/schedules/CAT/overrides', (req, res) => {
      res.status(422).send({ errors: [{ error: 'invalid service id' }] });
    });

    app.post('/users', (req, res) => {
      assert.equal(req.url, '/users', 'expected req.url');
      assert.ok(req.headers.accept && req.headers.authorization && req.headers.from, 'combined default & user-provided headers');
      res.status(201).send(createUser);
    });

    app.get('/escalation_policies/PPPPPPA', (req, res) => {
      assert.equal(req.url, '/escalation_policies/PPPPPPA?offset=0&total=true&limit=100', 'expected req.url');
      res.status(200).send(getEscalationPolicy);
    });

    app.get('/escalation_policies/PPPPPPA/log_entries', (req, res) => {
      assert.equal(req.url, '/escalation_policies/PPPPPPA/log_entries?offset=0&total=true&limit=100', 'expected req.url');
      res.status(200).send(listIncidentLogEntries);
    });

    app.get('/escalation_policies/PPPPPPA/notes', (req, res) => {
      assert.equal(req.url, '/escalation_policies/PPPPPPA/notes?offset=0&total=true&limit=100', 'expected req.url');
      res.status(200).send(listIncidentNotes);
    });

    app.get('/incidents', (req, res) => {
      assert.equal(req.url, '/incidents?since=2017-04-04T13%3A00%3A00Z&time_zone=UTC&offset=0&total=true&limit=100', 'expected req.url');
      res.status(200).send(listIncidents);
    });

    app.get('/schedules', (req, res) => {
      assert.equal(req.url, '/schedules?query=cat&offset=0&total=true&limit=100', 'expected req.url');
      res.status(200).send(listSchedules);
    });

    app.get('/users', (req, res) => {
      assert.equal(req.url, '/users?query=Dev%20Null&offset=0&total=true&limit=100', 'expected req.url');
      res.status(200).send(listUsers);
    });

    app.put('/users/PPPPPPA', (req, res) => {
      assert.equal(req.url, '/users/PPPPPPA', 'expected req.url');
      res.status(200).send(updateUser);
    });

    app.delete('/users/PPPPPPA', (req, res) => {
      assert.equal(req.url, '/users/PPPPPPA', 'expected req.url');
      res.status(204).send('');
    });

    app.get('/fakeDelay', (req, res) => {
      setTimeout((()=> res.status(200).send('')), 1000);
    });

    // 404 handler
    app.use((req, res) => {
      res.status(404).send({ error: 'Not Found' });
    });

    assert.end();
  });

  t.test('[pagerduty] start server', (assert) => {
    server = app.listen(3333, () => {
      assert.end();
    });
  });

  t.test('[pagerduty] [post] createOverride', (assert) => {
    pd.post({
      path: 'schedules/PPPPPPA/overrides',
      body: {
        override: {
          start: '2017-04-06T12:00:00+00:00',
          end: '2017-04-07T12:00:00+00:00',
          user: {
            id: 'PPPPPPA',
            type: 'user_reference',
            summary: 'Dev Null',
            self: 'https://api.pagerduty.com/users/PPPPPPA',
            html_url: 'https://company.pagerduty.com/users/PPPPPPA'
          }
        }
      }
    }, (err, res) => {
      assert.ifError(err, 'should not error');
      assert.deepEqual(res.body, createOverride);
      assert.end();
    });
  });

  t.test('[pagerduty] [post] createOverride error', (assert) => {
    pd.post({
      path: 'schedules/CAT/overrides',
      body: {
        override: {
          start: '2017-04-06T12:00:00+00:00',
          end: '2017-04-07T12:00:00+00:00',
          user: {
            id: 'CAT',
            type: 'user_reference',
            summary: 'Dev Null',
            self: 'https://api.pagerduty.com/users/PPPPPPA',
            html_url: 'https://company.pagerduty.com/users/PPPPPPA'
          }
        }
      }
    }, (err, res) => {
      assert.equal(err, 'HTTP status code 422: {"errors":[{"error":"invalid service id"}]}', 'expected error returned');
      assert.equal(res, undefined, 'only error object returned');
      assert.end();
    });
  });

  t.test('[pagerduty] [post] createUser', (assert) => {
    pd.post({
      path: 'users',
      body: {
        user: {
          type: 'user',
          name: 'Dev Null',
          email: 'devnull@company.com',
          time_zone: 'America/New_York',
          color: 'blue-violet',
          role: 'user'
        }
      },
      headers: {
        From: 'anotherdevnull@company.com'
      }
    }, (err, res) => {
      assert.ifError(err, 'should not error');
      assert.deepEqual(res.body, createUser);
      assert.end();
    });
  });

  t.test('[pagerduty] [get] getEscalationPolicy', (assert) => {
    pd.get({
      key: 'escalation_policy',
      path: 'escalation_policies/PPPPPPA'
    }, (err, res) => {
      assert.ifError(err, 'should not error');
      assert.deepEqual(res, [getEscalationPolicy.escalation_policy], 'expected pd.get response');
      assert.end();
    });
  });

  t.test('[pagerduty] [get] listIncidentLogEntries', (assert) => {
    pd.get({
      key: 'log_entries',
      path: 'escalation_policies/PPPPPPA/log_entries'
    }, (err, res) => {
      assert.ifError(err, 'should not error');
      assert.deepEqual(res, listIncidentLogEntries.log_entries, 'expected pd.get response');
      assert.end();
    });
  });

  t.test('[pagerduty] [get] listIncidentNotes', (assert) => {
    pd.get({
      key: 'notes',
      path: 'escalation_policies/PPPPPPA/notes'
    }, (err, res) => {
      assert.ifError(err, 'should not error');
      assert.deepEqual(res, listIncidentNotes.notes, 'expected pd.get response');
      assert.end();
    });
  });

  t.test('[pagerduty] [get] listIncidents', (assert) => {
    pd.get({
      key: 'incidents',
      path: 'incidents?since=2017-04-04T13%3A00%3A00Z&time_zone=UTC'
    }, (err, res) => {
      assert.ifError(err, 'should not error');
      assert.deepEqual(res, listIncidents.incidents, 'expected pd.get response');
      assert.end();
    });
  });

  t.test('[pagerduty] [get] listSchedules', (assert) => {
    pd.get({
      key: 'schedules',
      path: 'schedules?query=cat'
    }, (err, res) => {
      assert.ifError(err, 'should not error');
      assert.deepEqual(res, listSchedules.schedules, 'expected pd.get response');
      assert.end();
    });
  });

  t.test('[pagerduty] [get] listUsers', (assert) => {
    pd.get({
      key: 'users',
      path: 'users?query=Dev%20Null'
    }, (err, res) => {
      assert.ifError(err, 'should not error');
      assert.deepEqual(res, listUsers.users, 'expected pd.get response');
      assert.end();
    });
  });

  t.test('[pagerduty] [get] listUsers, no key', (assert) => {
    pd.get({
      path: 'users?query=Dev%20Null'
    }, (err, res) => {
      assert.ifError(err, 'should not error');
      assert.deepEqual(res, [listUsers], 'expected pd.get response');
      assert.end();
    });
  });

  t.test('[pagerduty] [get] handles 404 error', (assert) => {
    pd.get({
      path: 'undefined'
    }, (err, res) => {
      assert.equal(err, 'HTTP status code 404: {"error":"Not Found"}');
      assert.equal(res, undefined, 'only error object returned');
      assert.end();
    });
  });

  t.test('[pagerduty] [put] updateUsers', (assert) => {
    pd.put({
      path: 'users/PPPPPPA',
      body: {
        user: {
          description: 'Software Engineer'
        }
      }
    }, (err, res) => {
      assert.ifError(err, 'should not error');
      assert.deepEqual(res.body, updateUser, 'expected pd.put response');
      assert.end();
    });
  });

  t.test('[pagerduty] [delete] deleteUser delete', (assert) => {
    pd.delete({
      path: 'users/PPPPPPA'
    }, (err, res) => {
      assert.ifError(err, 'should not error');
      assert.deepEqual(res.body, '');
      assert.end();
    });
  });

  t.test('[pagerduty] timeout method option', (assert) => {
    pd.get({
      path: 'fakeDelay',
      timeout: 1
    }, (err) => {
      assert.equal(err.code, 'ETIMEDOUT');
      assert.end();
    });
  });

  t.test('[pagerduty] timeout client option', (assert) => {
    const pd1 = new PagerDuty({ pagerDutyToken: 'fakeaccesstoken', timeout: 1 });
    pd1.get({
      path: 'fakeDelay'
    }, (err) => {
      assert.equal(err.code, 'ETIMEDOUT');
      assert.end();
    });
  });

  t.test('[pagerduty] cleanup', (assert) => {
    server.close();
    assert.end();
  });

  t.end();
});


lib/utils.js
'use strict';

const querystring = require('querystring');
const url = require('url');

/**
 * A function that updates the URL for a GET request. This function checks if `offset`,
 * `total`, or `limit` have been provided in the querystring. If so, it uses the
 * provided values. If not, it uses default values.
 *
 * @name updateUrl
 * @param {string} address - The URL for a GET request.
 * @param {string} offset - The current offset value for response pagination.
 */
module.exports.updateUrl = updateUrl;
function updateUrl(address, offset) {
  const parsed = url.parse(address);
  const query = parsed.query;
  const qs = querystring.parse(query);

  if (!Object.hasOwnProperty.call(qs, 'offset')) qs.offset = offset;
  if (!Object.hasOwnProperty.call(qs, 'total')) qs.total = true;
  if (!Object.hasOwnProperty.call(qs, 'limit')) qs.limit = 100;

  parsed.search = `?${querystring.stringify(qs)}`;
  const formatted = url.format(parsed);
  return formatted;
}


lib/pagerduty.js
'use strict';
const baseUrl = process.env.TEST ? 'http://localhost:3333/' : 'https://api.pagerduty.com/';
const request = require('request');
const utils = require('./utils');

class PagerDuty {
  /**
  * Create a new PagerDuty API client.
  *
  * @param {string} [options.pagerDutyToken] - PagerDuty access token
  * @param {string} [options.timeout=10000] - number of milliseconds used for read and connection timouets
  */
  constructor (options) {
    this.token = options['pagerDutyToken'];
    this.timeout = Object.prototype.hasOwnProperty.call(options, 'timeout') ? options.timeout : 10000;
  }

  /**
   * Make a GET request to the PagerDuty API. You can choose to receive a specific
   * response body property by specifying an `options.key`; otherwise, the entire
   * response body will be returned. The PagerDuty access token can be provided
   * explicitly.
   *
   * @name get
   * @param {string} options.path - URL path and query
   * @param {string} [options.key] - Response body property to return
   * @param {string} [options.token] - PagerDuty access token
   * @param {string} [options.timeout] - number of milliseconds used for read and connection timouets
   *
   * @example
   * const PagerDuty = require('@mapbox/pagerduty');
   * const pd = new PagerDuty();
   * const id = 'PPPPPPA';
   * pd.get({
   *   path: `users/${id}`,
   *   key: 'user'
   * }, (err, res) => {
   *   if (err) throw err;
   *   console.log(`Successfully fetched ${res[0].name} with ID ${id}.`);
   * });
   */
  get (options, callback) {
    if (!options.path) return callback(new Error('Must provide options.path'));
    if (!options['token'] && !this['token']) return callback(new Error('Must provide PagerDuty access token'));
    if (!options['token']) options.token = this.token;
    if (!Object.prototype.hasOwnProperty.call(options, 'timeout')) options.timeout = this.timeout;

    let results = [];
    query(0);

    function query(offset) {
      request({
        method: 'GET',
        url: utils.updateUrl(`${baseUrl}${options.path}`, offset),
        json: true,
        headers: {
          Accept: 'application/vnd.pagerduty+json;version=2',
          Authorization: `Token token=${options.token}`
        },
        timeout: options.timeout
      }, (err, res, body) => {
        if (err) return callback(err);
        if (res.statusCode !== 200) return callback(`HTTP status code ${res.statusCode}: ${JSON.stringify(res.body)}`);

        results = options.key ? results.concat(body[options.key]) : results.concat(body);
        if (body.more) {
          offset += body.limit;
          query(offset);
        } else {
          return callback(null, results);
        }
      });
    }
  }

  /**
   * Make a POST request to the PagerDuty API. The PagerDuty access token can be
   * provided explicitly.
   *
   * @name post
   * @param {string} options.path - URL path and query
   * @param {object} options.body - POST request body
   * @param {object} [options.headers] - Additional headers, excluding `Accept`
   * and `Authorization`
   * @param {string} [options.token] - PagerDuty access token
   * @param {string} [options.timeout] - number of milliseconds used for read and connection timouets
   *
   * @example
   * const PagerDuty = require('@mapbox/pagerduty');
   * const pd = new PagerDuty();
   * pd.post({
   *   path: 'users',
   *   body: {
   *     user: {
   *       type: 'user',
   *       name: 'Dev Null',
   *       email: 'devnull@company.com',
   *       time_zone: 'America/New_York',
   *       color: 'blue-violet',
   *       role: 'user'
   *     }
   *   },
   *   headers: {
   *     From: 'anotherdevnull@company.com'
   *   }
   * }, (err, res) => {
   *   if (err) throw err;
   *   console.log(`Created ${res.body.user.summary} with ID ${res.body.user.id}.`);
   * });
   */
  post (options, callback) {
    if (!options.path) return callback(new Error('Must provide options.path'));
    if (!options.body) return callback(new Error('Must provide options.body'));
    if (!options['token'] && !this['token']) return callback(new Error('Must provide PagerDuty access token'));
    if (!options['token']) options.token = this.token;
    if (!Object.prototype.hasOwnProperty.call(options, 'timeout')) options.timeout = this.timeout;

    let headers = {
      Accept: 'application/vnd.pagerduty+json;version=2',
      Authorization: `Token token=${options.token}`
    };
    headers = Object.assign(headers, options.headers);

    request({
      method: 'POST',
      url: `${baseUrl}${options.path}`,
      headers: headers,
      json: options.body,
      timeout: options.timeout
    }, (err, res) => {
      if (err) return callback(err);
      if (res.statusCode !== 201) return callback(`HTTP status code ${res.statusCode}: ${JSON.stringify(res.body)}`);
      return callback(null, res);
    });
  }

  /**
   * Make a PUT request to the PagerDuty API. The PagerDuty access token can be
   * provided explicitly.
   *
   * @name put
   * @param {string} options.path - URL path and query
   * @param {object} options.body - PUT request body
   * @param {object} [options.headers] - Additional headers, excluding `Accept`
   * and `Authorization`
   * @param {string} [options.token] - PagerDuty access token
   * @param {string} [options.timeout] - number of milliseconds used for read and connection timouets
   *
   * @example
   * const PagerDuty = require('@mapbox/pagerduty');
   * const pd = new PagerDuty();
   * const id = 'PPPPPPA';
   * pd.put({
   *   path: `users/${id}`,
   *   body: {
   *     user: {
   *       job_title: 'Software Engineer'
   *     }
   *   }
   * }, (err, res) => {
   *   if (err) throw err;
   *   console.log(`Successfully updated ${res.body.user.name}'s user description to ${res.body.user.description}`);
   * });
   */
  put (options, callback) {
    if (!options.path) return callback(new Error('Must provide options.path'));
    if (!options.body) return callback(new Error('Must provide options.body'));
    if (!options['token'] && !this['token']) return callback(new Error('Must provide PagerDuty access token'));
    if (!options['token']) options.token = this.token;
    if (!Object.prototype.hasOwnProperty.call(options, 'timeout')) options.timeout = this.timeout;

    let headers = {
      Accept: 'application/vnd.pagerduty+json;version=2',
      Authorization: `Token token=${options.token}`
    };
    headers = Object.assign(headers, options.headers);

    request({
      method: 'PUT',
      url: `${baseUrl}${options.path}`,
      headers: headers,
      json: options.body,
      timeout: options.timeout
    }, (err, res) => {
      if (err) return callback(err);
      if (res.statusCode !== 200) return callback(`HTTP status code ${res.statusCode}: ${JSON.stringify(res.body)}`);
      return callback(null, res);
    });
  }

  /**
   * Make a DELETE request to the PagerDuty API. The PagerDuty access token can
   * be provided explicitly.
   *
   * @name delete
   * @param {string} options.path - URL path and query
   * @param {string} [options.token] - PagerDuty access token
   * @param {string} [options.timeout] - number of milliseconds used for read and connection timouets
   *
   * @example
   * const PagerDuty = require('@mapbox/pagerduty');
   * const pd = new PagerDuty();
   * const id = 'PPPPPPA';
   * pd.delete({
   *   path: `users/${id}`
   * }, (err, res) => {
   *   if (err) throw err;
   *   console.log(`Successfully deleted user ID ${id}.`);
   * });
   */
  delete (options, callback) {
    if (!options.path) return callback(new Error('Must provide options.path'));
    if (!options['token'] && !this['token']) return callback(new Error('Must provide PagerDuty access token'));
    if (!options['token']) options.token = this.token;
    if (!Object.prototype.hasOwnProperty.call(options, 'timeout')) options.timeout = this.timeout;

    request({
      method: 'DELETE',
      url: `${baseUrl}${options.path}`,
      headers: {
        Accept: 'application/vnd.pagerduty+json;version=2',
        Authorization: `Token token=${options.token}`
      },
      timeout: options.timeout
    }, (err, res) => {
      if (err) return callback(err);
      if (res.statusCode !== 204) return callback(`HTTP status code ${res.statusCode}: ${JSON.stringify(res.body)}`);
      return callback(null, res);
    });
  }
}

module.exports = PagerDuty;
